File: problems/2 - Add Two Numbers/Solution.java

// Definition for singly-linked list.
 class ListNode {
     int val;
     ListNode next;
     ListNode() {}
     ListNode(int val) { this.val = val; }
     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 }
 

public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode l3,l4=l3=null;
        ListNode a,b;
        a=l1;
        b=l2;
        int carry=0;
        for(;a!=null || b!=null;)
        {
            // if(carry==0)
            // {
                int sum =carry;
                if(a!=null){
                 sum+=a.val;
                 a=a.next;
                }
                 if(b!=null){
                 sum+=b.val;
                 b=b.next;
                 }
            int temp=sum%10;
            ListNode t1= new ListNode(temp);
            if(l3==null)
            l4=l3=t1;
            else 
            {
                l3.next=t1;
                l3= l3.next;
            }
            carry=sum/10;
           
        }
        if (carry!=0)
        {
            ListNode t1= new ListNode(carry);
            l3.next=t1;
        }
        return l4;
    }
}

File: problems/2 - Add Two Numbers/Solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        """
        Adds two numbers represented as linked lists.

        Args:
            l1 (ListNode): The first number.
            l2 (ListNode): The second number.

        Returns:
            ListNode: The sum of the two numbers.
        """
        dummy = ListNode()
        current = dummy
        carry = 0

        while l1 or l2 or carry:
            # Calculate the sum of the current nodes and the carry
            total = carry
            if l1:
                total += l1.val
                l1 = l1.next
            if l2:
                total += l2.val
                l2 = l2.next

            # Update the carry and create a new node with the digit of the sum
            current.next = ListNode(total % 10)
            current = current.next
            carry = total // 10

        return dummy.next
